[CIT-527]Software Management user-facing document (#409)
thin-edge/thin-edge.io: docs/src/tutorials/software-management.md

# Introduction

> This document describes how to install and use the **Software management** feature in thin-edge.

Author: fixed

> > Note: As of now the software management supported only on the **debian** based distros(Ex: Ubuntu, Debian),

Author: fixed

 which use the **apt**.

## Download and Install software management packages on the device

As a prequisite, install [tedge and tedge_mapper](../howto-guides/002_installation.md) if not installed already. 

The thin-edge software management packages are in repository on GitHub: [thin-edge.io](https://github.com/thin-edge/thin-edge.io/releases).

To download the package from github repository use the following command (use desired version):

```shell
curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/<package>_<version>_<arch>.deb
```

where:
> `version` -> thin-edge.io software management components version in x.x.x format
>
> `arch` -> architecture type (amd64, armhf)

> Download `tedg_apt_plugin` and `tedge_agent`

Author: fixed

```shell

> curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/0.1.1/tedge_apt_plugin_0.1.1_amd64.deb

Author: fixed

> curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/0.2.4/tedge_agent_0.2.4_amd64.deb

Author: fixed

```

Now the packages are downloaded, proceed to installation.
> Note: Some OSes may require you to use `sudo` to install packages and therefore all following commands may need `sudo`.

To install `tedge_apt_plugin` and `tedge_agent` on thin-edge device do:

```shell
dpkg -i tedge_apt_plugin_<version>_<arch>.deb
dpkg -i tedge_agent<version>_<arch>.deb
```

## Start and enable the software management services

The `tedge connect` will automatically start and enable the software management services.
Find more about [how to connect thin-edge device to cloud](../howto-guides/004_connect.md)

Once the thin-edge device is successfully connected to Cumulocity cloud, the **Software** option will be enabled and

> the list of softwares that are presently installed on the device will be visible as shown in the figure below.

Author: fixed

![Add new software](./images/start-software-management.png)

> Note: Disconnecting thin-edge device from cloud will stop and disable the software management services.

## Managing the device software on Cumulocity

More info about managing the device software can be found [here](https://cumulocity.com/guides/users-guide/device-management/#managing-device-software)

### Adding new software into software repository

1. In the Software repository page, click Add software at the right of the top menu bar.
2. In the resulting dialog box,
   - to add a new software, enter a name for the software (and confirm it by clicking Create new in the resulting window),
     a description and its version.
   - to add a new version, select the software for which you want to add a new version from the dropdown list in the Software
     field and enter a version and plugin type as **version::apt** (In the picture below the version is empty and are optional)
3. Optionally, you can define the device type filter when adding a new software.
4. Either upload a binary from the file system or specify a URL from where the software can be downloaded. As of now thin-edge does not

>    support both the options. So, select the **Provide a file path file path** option and give an **empty space**.

Author: fixed

5. Click Save.

Follow above mentioned steps as shown in the diagram

 ![Add new software](./images/add-new-software-to-repo.png)

> ### Deleting software or software version

Reviewer: Isnt that a duplicate with what is on line 92 "Remove software ...."

Author: This is about deleting from the repository.

Find more information about how to delete the software or the specific software version [here](https://cumulocity.com/guides/users-guide/device-management/#deleting-softwares-or-software-versions)

## Managing software on a device

Find more information about [how to manage the software](https://cumulocity.com/guides/users-guide/device-management/#managing-software-on-a-device) on a device.

### Install software on a device
Find about how to [Install a specific software on a device](https://cumulocity.com/guides/users-guide/device-management/#to-install-software-on-a-device)
> Note: Software profiles are not supported as of now on thin-edge.

> ### Update the software on a device

Reviewer: ```suggestion
### Update software on a device
```

Author: done

Find about how to [Update the software on a device](https://cumulocity.com/guides/users-guide/device-management/#to-update-software-on-a-device)

> ### Remove the software on a device

Reviewer: ```suggestion
### Remove software on a device
```

Author: done

# Introduction

> This document describes how to install and use the **Software management** feature in thin-edge.io.

Reviewer 2: Not sure why "Software management" needs to start with a capital 'S'. Just "software management" should be fine.

> > Note: As of now the software management only supports **debian** based distros(Ex: Ubuntu, Debian),

Reviewer 2: In that examples I think we need to highlight RaspberryPi OS more than Ubuntu and Debian as that's the primary target.

 which use the **apt**.

> ## Download and Install software management packages on the device

Reviewer 2: ```suggestion
## Download and install software management packages on the device
```

> As a prequisite, install [tedge and tedge_mapper](../howto-guides/002_installation.md) if not installed already. 

Reviewer 2: ```suggestion
As a prerequisite, install [tedge and tedge_mapper](../howto-guides/002_installation.md) if not installed already.
```

The thin-edge software management packages are in repository on GitHub: [thin-edge.io](https://github.com/thin-edge/thin-edge.io/releases).

To download the package from github repository use the following command (use desired version):

```shell
curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/<package>_<version>_<arch>.deb
```

where:
> `version` -> thin-edge.io software management components version in x.x.x format
>
> `arch` -> architecture type (amd64, armhf)

Download `tedge_apt_plugin` and `tedge_agent`

```shell
curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/0.3.0/tedge_apt_plugin_0.3.0_amd64.deb
curl -LJO https://github.com/thin-edge/thin-edge.io/releases/download/0.3.0/tedge_agent_0.3.0_amd64.deb
```

> Now the packages are downloaded, proceed to installation.

Reviewer 2: ```suggestion
Once the packages are downloaded, proceed to installation.
```

> Note: Some OSes may require you to use `sudo` to install packages and therefore all following commands may need `sudo`.

To install `tedge_apt_plugin` and `tedge_agent` on thin-edge device do:

```shell
dpkg -i tedge_apt_plugin_<version>_<arch>.deb
dpkg -i tedge_agent<version>_<arch>.deb
```

> ## Start and enable the software management services

Reviewer 2: ```suggestion
## Start and enable the software management feature
```
For the end-user it's the feature. Services are implementation details.

> The `tedge connect` will automatically start and enable the software management services.

Reviewer 2: ```suggestion
The `tedge connect` will automatically enable the software management feature.
```

Find more about [how to connect thin-edge device to cloud](../howto-guides/004_connect.md)

Once the thin-edge device is successfully connected to Cumulocity cloud, the **Software** option will be enabled and
the list of softwares that are installed on the device will be visible as shown in the figure below.

![Add new software](./images/start-software-management.png)

> > Note: Disconnecting thin-edge device from cloud will stop and disable the software management services.

Reviewer 2: ```suggestion
> Note: Disconnecting thin-edge device from cloud with `tedge disconnect` command will disable the software management feature as well.
```

Otherwise the the word "disconnect" might be confusing as it can also mean "disconnected from the network".

> ## Managing the device software on Cumulocity

Reviewer 2: ```suggestion
## Managing the device software from Cumulocity cloud
```

More info about managing the device software can be found [here](https://cumulocity.com/guides/users-guide/device-management/#managing-device-software)

### Adding new software into software repository

> 1. In the Software repository page, click Add software at the right of the top menu bar.

Reviewer 2: It would be better if we specify how to get to the software repository page from here. You could highlight that option on the left hand side pane in the UI in the image that you've linked below.

2. In the resulting dialog box,
   - to add a new software, enter a name for the software (and confirm it by clicking Create new in the resulting window),
     a description and its version.
   - to add a new version, select the software for which you want to add a new version from the dropdown list in the Software

>      field and enter a version and plugin type as **version::apt** (In the picture below the version is empty and are optional)

Reviewer 2: Why talk about the plugin type here as we currently only support one plugin and the type field is completely optional. We don't even need to mention it until we have a second plugin.

What we **must** highlight is that the version number is optional and can be left empty (white space) if the the user wants the latest version of that software.

3. Optionally, you can define the device type filter when adding a new software.

> 4. As of now thin-edge.io only supports software from apt repositories. So, select the **Provide a file path file path** option and give an **empty space**.

Reviewer 2: ```suggestion
4. As of now thin-edge.io only supports software from apt repositories. So, select the **Provide a file path** option and give an **empty space**.
```

5. Click Save.

Follow above mentioned steps as shown in the diagram

>  ![Add new software](./images/add-new-software-to-repo.png)

Reviewer 2: I'd suggest to move the image up and the step by step instructions follow, as a quick look at that image is pretty self explanatory.

Author: Followed the Cumulocity document :)

### Deleting software or software version

Find more information about how to delete the software or the specific software version [here](https://cumulocity.com/guides/users-guide/device-management/#deleting-softwares-or-software-versions)

## Managing software on a device

Find more information about [how to manage the software](https://cumulocity.com/guides/users-guide/device-management/#managing-software-on-a-device) on a device.

> 

Reviewer 2: You could add that image that I shared with you here which kinda gives a summary of how to use that UI.

### Install software on a device
Find about how to [Install a specific software on a device](https://cumulocity.com/guides/users-guide/device-management/#to-install-software-on-a-device)
> Note: Software profiles are not supported as of now on thin-edge.

### Update software on a device

> Find about how to [Update the software on a device](https://cumulocity.com/guides/users-guide/device-management/#to-update-software-on-a-device)

Reviewer 3: It's good to have external links to c8y. However, I would expect more information from this documentation.

Notably, what is the difference between manage/install/update?

### Remove software on a device
Hover over the software entry which you want to remove and click the **Remove** software icon and click on `Apply changes` button.

## Starting and Stopping the Software management services manually

> For debugging purpose the services can be started/stopped manually as shown below.

Reviewer 2: Not just for debugging purposes, but if someone wants to disable the software management feature and re-enable it later, then also these commands would be useful.
